text/0000-dynamic-query-params.md

```diff
@@ -0,0 +1,84 @@
+---
+stage: Proposed
+start-date: 2023-03-31T00:00:00.000Z
+release-date: # FIXME
+release-versions: # FIXME
+teams:
+  - framework
+prs: # FIXME
+project-link: # FIXME
+meta: # FIXME
+---
+
+# Setting dynamic query params in Controllers
+
+## Summary
+
+The current way developers bind query params to the Ember app is by defining the `queryParams` property on the Controller with hard coded values. See Ember docs on [Specifying Query Parameters](https://guides.emberjs.com/release/routing/query-params/#toc_specifying-query-parameters) section. It will be useful for developers who leverage a server-driven UI to dynamically bind query params to the Ember app.
+
+## Motivation
+
+With the current implementation of how query params work inside of an Ember app, developers are forced to have pre-knowledge of what query params are able to be binded to the app and hard code these values in the Controller. This may work for an app that is rigid in its interface, however there are use-cases where a server-driven UI can define the query params dynamically based on the user profile.
+
+Take for instance a simple filter bar where the filter options are known in advance. Hard coding values can work for this use-case because there is a limited amount of options and these options are generic for all users.
+
+```js
+// controller/simple/filter-bar.js
+queryParams = ['keywords', 'location', 'company']
+```
+
+Now let's expand this use-case and dynamically generate the simple filter bar into a complex filter, where the filter options are server-driven and tailored specifically to each user. The filter options are not known in advance and it will currently break how query params work in an Ember app.
+
+```js
+// controller/complex/filter-bar.js
+
+// this will not work in Ember today
+queryParams = this.model.filterOptions
+```
+
+## Detailed design
+
+I propose that we enable Ember the option to dynamically bind query params based on server-driven requirements.
+
+* Allow the Route's `setupController` to dynamically bind query params to the Controller's `queryParams` field
+* Keep the current option to statically bind query params to the Controller's `queryParams` field
+
+Example:
+
+```js
+// route/complex/filter-bar.js
+setupController(controller, model) {
```

Comment:
Why use any API at all?

https://github.com/emberjs/rfcs/pull/715

Could be enabled via optional-feature.json?

Personally, i'd like complexity removed from routes and controllers<3

Reply:
Oh yes that would be even better :) 

Query params should just be available and maintained (unless explicitly removed) for all transitions like how a vanilla web page links to other pages with query params